lib/phoenix/endpoint/instrument.ex

```diff
@@ -0,0 +1,187 @@
+defmodule Phoenix.Endpoint.Instrument do
+  @moduledoc false
+
+  # This is the arity that event callbacks in the instrumenter modules must
+  # have.
+  @event_callback_arity 3
+
+  def definstrument(otp_app, endpoint) do
+    app_instrumenters = app_instrumenters(otp_app, endpoint)
+
+    quote bind_quoted: [app_instrumenters: app_instrumenters] do
+      @doc """
+      Instruments the given function.
+
+      `event` is the event identifier (usually an atom) that specifies which
+      instrumenting function to call in the instrumenter modules. `runtime` is
+      metadata to be associated with the event at runtime (e.g., the query being
+      issued if the event to instrument is a DB query).
+
+      ## Examples
+
+          instrument :render_view, [view: "index.html"], fn ->
+            render conn, "index.html"
+          end
+
+      """
+      defmacro instrument(event, runtime \\ nil, fun) do
+        compile = Macro.escape(Phoenix.Endpoint.Instrument.strip_caller(__CALLER__))
+
+        quote do
+          unquote(__MODULE__).instrument(
+            unquote(event),
+            unquote(compile),
+            unquote(runtime),
+            unquote(fun)
+          )
+        end
+      end
+
+      # For each event in any of the instrumenters, we must generate a
+      # clause of the `instrument/4` function. It'll look like this:
+      #
+      #   def instrument(:my_event, compile, runtime, fun) do
+      #     res0 = Inst0.my_event(:start, compile, runtime)
+      #     ...
+      #
+      #     start = current_time()
+      #     try do
+      #       fun.()
+      #     after
+      #       diff = ...
+      #       Inst0.my_event(:stop, diff, res0)
+      #       ...
+      #     end
+      #   end
+      #
+      @doc false
+      def instrument(event, compile, runtime, fun)
+
+      for {event, instrumenters} <- app_instrumenters do
+        def instrument(unquote(event), var!(compile), var!(runtime), fun)
+            when is_map(var!(compile)) and is_function(fun, 0) do
+          unquote(Phoenix.Endpoint.Instrument.compile_start_callbacks(event, instrumenters))
+          start = current_time()
+          try do
+            fun.()
+          after
+            var!(diff) = time_diff(start, current_time())
+            unquote(Phoenix.Endpoint.Instrument.compile_stop_callbacks(event, instrumenters))
+          end
+        end
+      end
+
+      # Catch-all clause
+      def instrument(_event, compile, _runtime, fun)
+          when is_map(compile) and is_function(fun, 0) do
+        fun.()
+      end
+
+      # Let's define the current_time/0 and time_diff/2 functions based on the
+      # existence of :erlang.monotonic_time/1.
```

Comment:
Let's leave a todo: `# TODO: Remove this once Phoenix supports only Elixir 1.2`.